docs/code-organisation-guidelines.md

```diff
@@ -0,0 +1,65 @@
+# Code Organisation Guidelines
+
+The document specifies how code in the Cardano SL git repository should be organised. Consistent
+module naming schemes and file locations make the code relatively uniform so that it is easier for
+developers (internal and external), engineers working on formal specifications and auditors
+(internal and external) to navigate the code base.
+
+Currently, the code base does not conform to these guidelines but it is hoped that over time it
+can move from non-conforming to conforming. Making code conform to these guidelines in one big
+pull request or one long lived refactoring branch would be a huge amount of work and/or cause a
+lot of disruption to the regular flow of development. Instead, these guidelines should be applied
+as a series of small easy-to-review commits that can be merged relatively quickly.
+
+* The `cardano-sl` git repository includes a number of distinct Haskell packages like `core`, `util`,
+  `explorer`, `networking` and so on. All Haskell modules within each of these packages should be
+  given a unique name prefix. All modules within `core` would have a prefix `Pos.Core`, modules
+  within `util` would have a name prefix `Pos.Util`, modules within `networking` might have a prefix
+  of `Pos.Network` and so on. This means that when looking at the imports for a module, it is
+  immediately obvious which package each import comes from. As an example of a file that needs to be
+  renamed/moved the file `core/Pos/Aeson/Core.hs` should be renamed to `core/Pos/Core/Aeson.hs` so
+  that if it is imported in another module its name `Pos.Core.Aeson` would immediately suggest its
+  location in the source tree.
+
+* All library code for a Haskell packages should be in the `Pos` directory of that package.
+  Currently some packages use `src` and then `src` has both `Pos` and `Test` sub-directories. Since
+  all test code should be in the `test` sub directory (see below), we should choose `Pos` over
+  `src/Pos`. For example `lib` currently has testing related code in the `src` directory and if
+  that test code moves to the `test` directory then there is no need for `src`.
+
+* Tests for a Haskell package should always be a part of the same package as the implementation of
+  the code it is testing. Test code should not be included in the library that a package provides.
+  All test code for a Haskell package should be in the `test` directory of that package and modules
+  with testing related code should have module names staring with `Test` and then reflecting the
+  Haskell package which the code is testing, so test code for the `core` package would have all its
+  test module names begin with `Test.Pos.Core`. An example of where the current code base
+  contravenes these guidelines is that `core` and `txp` currently don't define any tests, but at
+  least some of the functionality in these two modules are tested in `lib` package.
```

Comment:
I don't have strong opinions on test naming conventions, but I do think we should have test naming conventions that enable autodiscovery of tests. As an example, `hspec-discover` expects to find modules matching the pattern `**/*Spec.hs`, and having `test/Test/Foo/BarSpec.hs` is redundant compared to `test/Foo/BarSpec.hs`. `tasty-discover` has it's own mechanism/convention for discovery.

Reply:
Auto discovery of tests is very much tied to the testing framework used. I've used Hspec in the past, but personally I would not choose it over Hedgehog.

Reply:
`hspec` is more of a library for test organization and delegates to `hunit` for actual testing, IIRC. It has integration points with `QuickCheck` and `hedgehog`. Does Hedgehog do non-property testing as well?

Reply:
Maybe? :) It is possible to non-property testing with Hedgehog using something like:
```
Hedgehog.withTests 1 $ Hedgehog.property prop_wibble
```

Reply:
Trying to somehow unify the two point of views (hopefully not creating a 3rd, conflicting one 😁 ):

1. We should probably pick one testing framework and stick with it. I know we all have preferences (for example I am a fan of `tasty` 😁 ) but hopefully after a collective decision we can write this down and forget about it. _Dura lex, sed lex_ Latins would say, that rought translates as "it's a hard law, but it's the law". No matter which framework we'll choose, somebody will be disappointed, but if that's the majority expressing consensus, at least the mouthful would be less sour to swallow 😛 

2. Regarding @parsonsmatt 's comment, I think the issue might be largely mitigated by which framework we end up using as part of (1.), but I will be provocative here: I tend to lean towards @erikd 's point of view that we shouldn't make compromises on the namespace just to please the autodiscover gods. Not only that, but do you guys find auto-discovery for tests an essential tool? I think that's handy, but in my opinion these mechanisms fall into the same trap of `autodiscover` when it comes to code navigation. If I have to explicitly mentions all the tests I want to run in my main module, it means I can probably have a holistic view of all the different things I'm testing, and easily jump to them via `ctags` or similar. Not to say auto-discovery is not useful, but I have done OK in the past also without. Am I a fool? :trollface:

Reply:
I've used hspec's autodiscovery and found it something that I could easily live without.

With Hedgehog, I use the file level test discovery, but when I add a new file of tests, I do it manually.

Reply:
I'm immensely forgetful and lazy, and I have strong preferences for things that I don't need to remember or do myself. I like `hspec` autodiscovery because it means that I just write tests and then they run -- I don't have to remember to add them to `main`, they just work automagically. For the same reason, I like `hpack`'s "autodiscovery" of exposed modules.

It's an ergonomic thing that has essentially 0 cost and a decent recurring benefit. If we adopt a convention that *doesn't* have an autodiscovery tool, I will probably write one for it :smile:

Reply:
I have to say, I'm always deeply suspicious of test auto-discovery, and magic generally.